### ISSUE TITLE
Overlay2 on ZFS file operations "invalid argument"

### ISSUE THREAD

**Issue by mjforan** [moby/moby#46640]
### Description

When Docker is installed on a ZFS file system it will attempt to use the Overlay2 storage driver by default. This causes issues with modifying certain files in containers.

This is reproducible in a fresh Ubuntu 22.04 VM. A default (ext4 instead of ZFS) Ubuntu VM does not exhibit this behavior. The issue can be resolved by wiping `/var/lib/docker` and switching to the ZFS storage driver.

### Reproduce

1. Install Ubuntu 22.04 with ZFS root file system
1. Upgrade all packages
1. Install docker from apt repository
1. `docker run --rm -it ros:noetic-robot`
1. `apt remove -y ros-noetic-robot`

```
dpkg: error processing package ros-noetic-robot (--remove):
 cannot remove '/usr/share/doc/ros-noetic-robot': Invalid argument
dpkg: too many errors, stopping
```

`rm -rf /usr/share/doc/ros-noetic-robot` removes the **contents** of the folder and returns without error, but the folder itself remains.
`rmdir` and `mv` on that folder fail with "invalid argument".
This problem is not exhibited with folders I create live from the shell, even though they have the exact same permissions.

### Expected behavior

No problems removing files, "the package was removed successfully", etc.

### docker version

```bash
Client: Docker Engine - Community
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:31:44 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:31:44 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.24
  GitCommit:        61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc:
  Version:          1.1.9
  GitCommit:        v1.1.9-0-gccaecfc
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client: Docker Engine - Community
 Version:    24.0.6
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.21.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 13
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: zfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 nvidia
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc version: v1.1.9-0-gccaecfc
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.2.0-34-generic
 Operating System: Ubuntu 22.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 31.05GiB
 Name: mattf
 ID: 12c3935d-a978-4a00-aae9-c75d6bc2b736
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  repo.codev.mitre.org
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

*For search engines*
I initially came across this issue when running `apt upgrade` with the following error:
```
Processing triggers for sgml-base (1.29.1) ...
cannot rename /var/lib/sgml-base/supercatalog to /var/lib/sgml-base/supercatalog.old: Invalid argument at /usr/sbin/update-catalog line 322.
dpkg: error processing package sgml-base (--configure):
installed sgml-base package post-installation script subprocess returned error exit status 22
```

**Comment by corhere** [comment]
This is a known issue: duplicate of #46337, will be fixed by #45890

**Comment by mjforan** [comment]
@corhere Thanks! I swear I searched for related issues but this one escaped my keywords :P
